src/Polysemy/Internal/Final.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE TemplateHaskell #-}
+module Polysemy.Internal.Final where
+
+import Polysemy
+import Polysemy.Internal
+
+-----------------------------------------------------------------------------
+-- | This represents a function which produces
+-- an action of the final monad @m@ given:
+--
+--   * The initial effectful state at the moment the action
+--     is to be executed.
+--
+--   * A way to convert @z@ (which is typically @'Sem' r@)to @m@ by
+--     threading the effectful state through.
+--
+--   * An inspector that is able to view some value within the
+--     effectful state if the effectful state contains any values.
+--
+-- A @'Polysemy.Internal.Union.Weaving'@ provides these components,
+-- hence the name 'ThroughWeavingToFinal'.
+type ThroughWeavingToFinal m z a =
+     forall f
+   . Functor f
+  => f ()
+  -> (forall x. f (z x) -> m (f x))
+  -> (forall x. f x -> Maybe x)
+  -> m (f a)
+
+-----------------------------------------------------------------------------
+-- | An effect for embedding higher-order actions in the final target monad
+-- of the effect stack.
+--
+-- This is very useful for writing interpreters that interpret higher-order
+-- effects in terms of the final monad.
+--
+-- 'Final' is slightly weaker, but using it is otherwise more preferable
+-- than using a provided lowering function (a la 'Polysemy.Async.lowerAsync'):
```

Comment:
Probably not. `Final` _is_ slightly weaker than a lowering function, so it's half-imaginable that people would create an interpreter that would need the extra power of a lowering function rather than `Final`.
Thing is, the restrictions placed by `Final` are already placed by the need to thread functorial state from previous interpreters through (for example, you can't do a `UnliftIO` interpreter even with a lowering function). I can't think of one example where an explicit lowering function would give you the power to do something `Final` can't.

Reply:
I removed the comment.